src/nbl/asset/utils/ISPIRVDebloater.cpp

```diff
@@ -0,0 +1,115 @@
+        ISPIRVOptimizer::EOP_ELIM_DEAD_VARIABLES,
+        ISPIRVOptimizer::EOP_ELIM_DEAD_CONSTANTS,
+        // TODO: remove unused type. no dedicated flag for type. maybe need agressive dce to remove type.
+        ISPIRVOptimizer::EOP_TRIM_CAPABILITIES,
+    };
+    m_optimizer = core::make_smart_refctd_ptr<ISPIRVOptimizer>(std::span(optimizationPasses));
+}
+
+ISPIRVDebloater::Result ISPIRVDebloater::debloat(const  ICPUBuffer* spirvBuffer, std::span<const EntryPoint> entryPoints, system::logger_opt_ptr logger) const
+{
+    const auto* spirv = static_cast<const uint32_t*>(spirvBuffer->getPointer());
+    const auto spirvDwordCount = spirvBuffer->getSize() / 4;
+
+    auto getHlslShaderStage = [](spv::ExecutionModel executionModel) -> hlsl::ShaderStage
+      {
+        switch (executionModel)
+        {
+            case  spv::ExecutionModelVertex : return hlsl::ESS_VERTEX;
+            case  spv::ExecutionModelTessellationControl : return hlsl::ESS_TESSELLATION_CONTROL;
+            case  spv::ExecutionModelTessellationEvaluation : return hlsl::ESS_TESSELLATION_EVALUATION;
+            case  spv::ExecutionModelGeometry : return hlsl::ESS_GEOMETRY;
+            case  spv::ExecutionModelFragment : return hlsl::ESS_FRAGMENT;
+            case  spv::ExecutionModelGLCompute : return hlsl::ESS_COMPUTE;
+            case  spv::ExecutionModelTaskEXT : return hlsl::ESS_TASK;
+            case  spv::ExecutionModelMeshEXT: return hlsl::ESS_MESH;
+            case  spv::ExecutionModelRayGenerationKHR : return hlsl::ESS_RAYGEN;
+            case  spv::ExecutionModelAnyHitKHR : return hlsl::ESS_ANY_HIT;
+            case  spv::ExecutionModelClosestHitKHR : return hlsl::ESS_CLOSEST_HIT;
+            case  spv::ExecutionModelMissKHR : return hlsl::ESS_MISS;
+            case  spv::ExecutionModelIntersectionKHR : return hlsl::ESS_INTERSECTION;
+            case  spv::ExecutionModelCallableKHR : return hlsl::ESS_CALLABLE;
+            default:
+            {
+                assert(false);
+                return hlsl::ESS_UNKNOWN;
+            }
+        }
+      };
+    static constexpr auto HEADER_SIZE = 5;
+
+    core::vector<bool> foundEntryPoints(entryPoints.size(), false);
+    std::vector<uint32_t> minimizedSpirv;
+    bool needDebloat = false;
+
+    auto offset = HEADER_SIZE;
+
+    while (offset < spirvDwordCount) {
+        const auto curOffset = offset;
+        const auto instruction = spirv[curOffset];
+
+        const auto length = instruction >> 16;
+        const auto opcode = instruction & 0x0ffffu;
+
+        offset += length;
+
+        if (opcode == spv::OpEntryPoint)
+        {
+            const auto curExecutionModel = static_cast<spv::ExecutionModel>(spirv[curOffset + 1]);
+
+            // TODO: check whether this reinterpret_cast is UB
+            const auto curEntryPointName = std::string_view(reinterpret_cast<const char*>(spirv + curOffset + 3)); // entryPoint name is the third parameter
```

Comment:
no its not UB, but will the string be null terminated as per SPIR-V ?

Also you need to protect against someone shoving garbage into the debloater, so please protect against non-null terminated strings. You need to actually find the null terminator with a limit of looking up to `COUNT-curOffset-4` away.

also `while` checks that `offset<COUNT` but it doesn't check `offset+3<COUNT`